src/metabase/pulse.clj

```diff
@@ -53,50 +108,48 @@
+   :day-of-week   (str timestamp)
+   :week-of-year  (str timestamp)
+   :month-of-year (str timestamp)
+   (f/unparse (f/formatter "MMM d, YYYY") (c/from-long timestamp))))
+
+(def ^:private year  (comp t/year  t/now))
+(def ^:private month (comp t/month t/now))
+(def ^:private day   (comp t/day   t/now))
+
+(defn- date->interval-name [date interval-start interval this-interval-name last-interval-name]
+  (cond
+    (t/within? (t/interval interval-start                    (t/plus interval-start interval)) date) this-interval-name
+    (t/within? (t/interval (t/minus interval-start interval) interval-start)                   date) last-interval-name))
+
+(defn- start-of-this-week    [] (-> (org.joda.time.LocalDate.) .weekOfWeekyear .roundFloorCopy .toDateTimeAtStartOfDay))
+(defn- start-of-this-quarter [] (t/date-midnight (year) (+ 1 (* 3 (Math/floor (/ (dec (month)) 3))))))
 
 (defn- format-timestamp-relative
   "Formats timestamps with relative names (today, yesterday, this *, last *) based on column :unit, if possible, otherwie returns nil"
-  [timestamp col]
-  (case (:unit col)
-    :day      (let [date    (c/from-long timestamp)
-                    d-start (t/date-midnight (t/year (t/now)) (t/month (t/now)) (t/day (t/now)))]
-                (cond (t/within? (t/interval d-start (t/plus d-start (t/days 1))) date) "Today"
-                      (t/within? (t/interval (t/minus d-start (t/days 1)) d-start) date) "Yesterday"))
-    :week     (let [date    (c/from-long timestamp)
-                    w-start (-> (new org.joda.time.LocalDate) .weekOfWeekyear .roundFloorCopy .toDateTimeAtStartOfDay)]
-                (cond (t/within? (t/interval w-start (t/plus w-start (t/weeks 1))) date) "This week"
-                      (t/within? (t/interval (t/minus w-start (t/weeks 1)) w-start) date) "Last week"))
-    :month    (let [date    (c/from-long timestamp)
-                    m-start  (t/date-midnight (t/year (t/now)) (t/month (t/now)))]
-                (cond (t/within? (t/interval m-start (t/plus m-start (t/months 1))) date) "This month"
-                      (t/within? (t/interval (t/minus m-start (t/months 1)) m-start) date) "Last month"))
-    :quarter  (let [date    (c/from-long timestamp)
-                    q-start (t/date-midnight (t/year (t/now)) (+ 1 (* 3 (Math/floor (/ (- (t/month (t/now)) 1) 3)))))]
-                (cond (t/within? (t/interval q-start (t/plus q-start (t/months 3))) date) "This quarter"
-                      (t/within? (t/interval (t/minus q-start (t/months 3)) q-start) date) "Last quarter"))
-    :year     (let [date    (t/date-midnight timestamp)
-                    y-start (t/date-midnight (t/year (t/now)))]
-                (cond (t/within? (t/interval y-start (t/plus y-start (t/years 1))) date) "This year"
-                      (t/within? (t/interval (t/minus y-start (t/years 1)) y-start) date) "Last year"))
-    nil))
```

Comment:
I found this function a little unwieldy, it was essentially repeating the same logic 5 times so I refactored that logic into a separate function.